jstests/aggregation/isodates.js

```diff
@@ -0,0 +1,243 @@
+/*
```

Comment:
I don't like reducing the coverage, as you already mentioned the algorithms aren't easy to follow, so I'd rather have an extensive test suite, which contains every edge case to prevent regression if someone changes this code.

[jstests/aggregation/bugs/server6189.js](https://github.com/mongodb/mongo/blob/r3.3.2/jstests/aggregation/bugs/server6189.js): ist for testing dates before 1970, I could try adding the new operators here. But I would still keep the other file for testing all edges there are.

[src/mongo/db/pipeline/expression_test.cpp](https://github.com/mongodb/mongo/blob/master/src/mongo/db/pipeline/expression_test.cpp): doesn't include any tests on dates (that I see). I was searching the code for date tests and only found the javascript ones. And also because I'm more familiar with javascript than with c++ I added  the new tests as new javascript file.